fem/pfespace.cpp

```diff
@@ -3507,6 +3507,127 @@ void ParFiniteElementSpace::Update(bool want_transform)
    }
 }
 
+void ParFiniteElementSpace::GetTrueUpdateOperator(OperatorHandle &T)
+{
+   bool need_P_old = false;
+   std::unique_ptr<HypreParMatrix> P_old;
+
+   T.Clear();  // preserves T.Type()
+
+   switch (mesh->GetLastOperation())
+   {
+      case Mesh::REFINE:
+      {
+         if (T.Type() == Operator::Hypre_ParCSR)
+         {
+            SetUpdateOperatorType(Operator::MFEM_SPARSEMAT);
+            need_P_old = true;
+         }
+         break;
+      }
+
+      case Mesh::DEREFINE:
+      {
+         if (T.Type() == Operator::Hypre_ParCSR)
+         {
+            need_P_old = true;
+         }
+         break;
+      }
+
+      case Mesh::REBALANCE:
+      {
+         if (T.Type() == Operator::Hypre_ParCSR)
+         {
+            need_P_old = true;
+         }
+         break;
+      }
+
+      default:
+         break;
+   }
+   if (need_P_old)
+   {
+      // Save P as P_old:
+      P_old.reset(Dof_TrueDof_Matrix()); // make sure P is constructed
+      P_old->CopyRowStarts();
+      P_old->CopyColStarts();
+      P = nullptr;
+   }
+
+   Update(true);
+
+   if (NURBSext)   // NURBS spaces are not supported yet!
+   {
+      return;
+   }
+
+   switch (mesh->GetLastOperation())
+   {
+      case Mesh::REFINE:
+      {
+         if (T.Type() == Operator::Hypre_ParCSR)
+         {
+            SparseMatrix *Tsm = Th.Is<SparseMatrix>();
+            MFEM_VERIFY(Tsm != nullptr,
+                        "invalid update (refine) operator type!");
+            // Form the product: R_new . Tsm . P_old:
+            SparseMatrix *R_T = mfem::Mult(*GetRestrictionMatrix(), *Tsm);
+            T.Reset(P_old->LeftDiagMult(*R_T, GetTrueDofOffsets()));
+            delete R_T;
```

Comment:
```suggestion
            ::std::unique_ptr<SparseMatrix> R_T = mfem::Mult(*GetRestrictionMatrix(), *Tsm);
            T.Reset(P_old->LeftDiagMult(*R_T, GetTrueDofOffsets()));
```

Reply:
Analogous for lines 3602-3604 and 3618-3621. I wouldn't have said anything but you're already using a `std::unique_ptr` above on line 3513.

Reply:
Done in 5e3974d9.